sympy/core/operations.py

```diff
@@ -504,3 +510,149 @@ def args(self):
+    ========
+
+    >>> from sympy import Add, Expr, Symbol
+    >>> from sympy.core.add import add
+
+    >>> class NewExpr(Expr):
+    ...     @staticmethod
+    ...     def _add_handler(*args, **kwargs):
+    ...         return NewAdd(*args, **kwargs)
+    >>> class NewAdd(NewExpr, Add):
+    ...     pass
+
+    >>> @add.register_priority(Expr, NewExpr)
+    ... @add.register_priority(NewExpr, NewExpr)
+    ... def _(_1, _2):
+    ...     return NewExpr
+
+    >>> a, b = Symbol('a'), NewExpr()
+    >>> add(a, b) == NewAdd(a, b)
+    True
+
+    Registered priority selector can return any type.
+
+    >>> class Foo(Expr):
+    ...     @staticmethod
+    ...     def _add_handler(*args, **kwargs):
+    ...         raise NotImplementedError
+    >>> class Bar(Expr):
+    ...     @staticmethod
+    ...     def _add_handler(*args, **kwargs):
+    ...         raise NotImplementedError
+    >>> class FooBar(Expr):
+    ...     @staticmethod
+    ...     def _add_handler(*args, **kwargs):
+    ...         return NewAdd(*args, **kwargs)
+
+    >>> @add.register_priority(Foo, Bar)
+    ... def _(_1, _2):
+    ...     return FooBar
+
+    >>> add(Foo(), Bar()) == NewAdd(Foo(), Bar())
+    True
+
+    """
+    def __init__(self, name, doc=None):
+        self.name = name
+        self.doc = doc
+        self.handlermethod = "_%s_handler" % name
+        self.handlergetter = attrgetter(self.handlermethod)
+        self._dispatcher = Dispatcher(name)
+
+    def __repr__(self):
+        return "<dispatched %s>" % self.name
+
+    def register_priority(self, cls1, cls2, on_ambiguity=ambiguity_register_error):
+        """
+        Register the priority between two types, in both straight and reversed order.
+        Registered function must return the class which has the handler that is to be used.
+        """
+        def _(func):
+            self._dispatcher.add((cls1, cls2), func, on_ambiguity=on_ambiguity)
+            self._dispatcher.add((cls2, cls1), func, on_ambiguity=on_ambiguity)
+            return func
+        return _
+
+    @cacheit
+    def __call__(self, *args, **kwargs):
+        if kwargs.get('sympify', True):
+            args = tuple(map(_sympify, args))
+        types = frozenset(map(type, args))
+
+        kwargs.update(sympify=False)
+        return self.dispatch(types)(*args, **kwargs)
+
+    @cacheit
+    def dispatch(self, types):
+        """
+        Select the type with highest priority, and return its handler.
+        """
+
+        # Quick exit for the case where all arguments are of same type
+        # or all handlermethods are same
+        handlers = set(map(self.handlergetter, types))
+        if len(handlers) == 1:
+            for h in handlers:
+                return h
+
+        # Sieve the types with low op_priority or no handler
+        # This makes it faster than using dispatcher on every types.
+        sieved_data = tied_max(
+            (typ for typ in types), key=lambda typ: getattr(typ, '_op_priority', 0)
```

Comment:
Should work as the shorter
```suggestion
            types, key=lambda typ: getattr(typ, '_op_priority', 0)
```

Reply:
This is the silliest mistake I've ever made.

Reply:
If that's the case, presumably we can eke out some performance with `operator.attrgetter` here too.

Reply:
They are all sympified here.

Reply:
Resolving the conversation.